zjit/src/cruby_methods.rs

```diff
@@ -59,10 +66,86 @@ fn annotate_c_method(props_map: &mut HashMap<*mut c_void, FnProperties>, class:
     props_map.insert(fn_ptr, props);
 }
 
+/// Look up a method and find its builtin function pointer by parsing its ISEQ
+fn annotate_builtin_method(props_map: &mut HashMap<*mut c_void, FnProperties>, class: VALUE, method_name: &'static str, mut props: FnProperties) {
+    unsafe {
+        let method_id = rb_intern2(method_name.as_ptr().cast(), method_name.len().try_into().unwrap());
+        let method = rb_method_entry_at(class, method_id);
+        if method.is_null() {
+            panic!("Method {}#{} not found", std::ffi::CStr::from_ptr(rb_class2name(class)).to_str().unwrap_or("?"), method_name);
+        }
+
+        // Cast ME to CME - they have identical layout
+        let cme = method.cast::<rb_callable_method_entry_t>();
+        let def_type = get_cme_def_type(cme);
+
+        if def_type != VM_METHOD_TYPE_ISEQ {
+            panic!("Method {}#{} is not an ISEQ method (type: {})",
+                std::ffi::CStr::from_ptr(rb_class2name(class)).to_str().unwrap_or("?"),
+                method_name, def_type);
+        }
+
+        // Get the ISEQ from the method definition
+        let iseq = get_def_iseq_ptr((*cme).def);
+        if iseq.is_null() {
+            panic!("Failed to get ISEQ for {}#{}",
+                std::ffi::CStr::from_ptr(rb_class2name(class)).to_str().unwrap_or("?"),
+                method_name);
+        }
+
+        // Check if the ISEQ has the BUILTIN_ATTR_LEAF attribute
+        let builtin_attrs = rb_zjit_get_iseq_body_builtin_attrs(iseq);
+        if (builtin_attrs & BUILTIN_ATTR_LEAF) != 0 {
+            // If the method is marked as leaf via Primitive.attr!, set leaf to true
+            props.leaf = true;
+            // Leaf methods don't trigger GC
+            props.no_gc = true;
```

Comment:
Could we avoid deriving `no_gc` from `leaf`, and just annotate `no_gc` separately in ZJIT for now? 

We have the history of adding `:no_gc` attribute while `:leaf` attribute already exists https://github.com/ruby/ruby/pull/7511, so they're considered separate things. `leaf` just means that it doesn't call other methods, so semantically it's still allowed to allocate.